File: java/Samples/Biometrics/CS/AbisSampleCS/SubjectEditor/BiometricPreviewPage.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Gui;
using Neurotec.Images;

namespace Neurotec.Samples
{
	public partial class BiometricPreviewPage : Neurotec.Samples.PageBase
	{
		#region Public constructor

		public BiometricPreviewPage()
		{
			InitializeComponent();
			DoubleBuffered = true;
		}

		#endregion

		#region Private fields

		private SubjectTreeControl.Node _node;
		private NBiometric _biometric;
		private object _view;

		#endregion

		#region Public methods

		public override void OnNavigatedTo(params object[] args)
		{
			if (args == null || args.Length != 1) throw new ArgumentException("args");

			bool showZoomControls = true;
			bool showBinarizedCheckbox = false;
			bool canSave = false;
			_view = null;
			_node = (SubjectTreeControl.Node)args[0];
			if (_node == null || !_node.IsBiometricNode) throw new ArgumentException("args");
			panelView.Controls.Clear();
			generalizeProgressView.Clear();
			generalizeProgressView.Visible = _node.IsGeneralizedNode;
			icaoWarningView.Visible = false;
			_biometric = _node.Items.First();
			switch (_node.BiometricType)
			{
				case NBiometricType.Finger:
				case NBiometricType.Palm:
					{
						NFrictionRidge first = _biometric as NFrictionRidge;
						NFingerView view = new NFingerView()
						{
							Dock = DockStyle.Fill,
							ForeColor = Color.Red,
							Finger = first
						};
						_view = view;
						panelView.Controls.Add(view);
						horizontalZoomSlider.View = view;
						showBinarizedCheckbox = first.BinarizedImage != null;
						canSave = first.Image != null;
						if (generalizeProgressView.Visible)
						{
							var generalized = _node.GetAllGeneralized();
							generalizeProgressView.View = view;
							generalizeProgressView.Biometrics = _node.Items;
							generalizeProgressView.Generalized = generalized;
							generalizeProgressView.Selected = generalized.FirstOrDefault() ?? first;
						}

						if (showBinarizedCheckbox && chbShowBinarized.Checked)
							view.ShownImage = ShownImage.Result;
						break;
					}
				case NBiometricType.Iris:
					{
						NIrisView view = new NIrisView()
						{
							Dock = DockStyle.Fill,
							Iris = _biometric as NIris
						};
						panelView.Controls.Add(view);
						horizontalZoomSlider.View = view;
						_view = view;
						canSave = view.Iris.Image != null;
						break;
					}
				case NBiometricType.Face:
					{
						NFace first = _biometric as NFace;
						NFaceView view = new NFaceView()
						{
							Dock = DockStyle.Fill,
							ShowIcaoArrows = false,
							Face = first
						};
						_view = view;
						panelView.Controls.Add(view);
						horizontalZoomSlider.View = view;
						canSave = view.Face.Image != null;
						if (_node.AllItems.Cast<NFace>().Any(WasIcaoCheckPerformed))
						{
							icaoWarningView.Visible = true;
						}
						if (generalizeProgressView.Visible)
						{
							var generalized = _node.GetAllGeneralized();
							generalizeProgressView.View = view;
							generalizeProgressView.IcaoView = icaoWarningView;
							generalizeProgressView.Biometrics = _node.Items;
							generalizeProgressView.Generalized = generalized;
							generalizeProgressView.Selected = generalized.FirstOrDefault() ?? first;
						}
						else
						{
							icaoWarningView.Face = first;
						}
						break;
					}
				case NBiometricType.Voice:
					{
						showZoomControls = false;
						VoiceView view = new VoiceView() { Voice = _biometric as NVoice };
						panelView.Controls.Add(view);
						_view = view;
						canSave = view.Voice.SoundBuffer != null;
						break;
					}
			};

			horizontalZoomSlider.Visible = showZoomControls;
			chbShowBinarized.Visible = showBinarizedCheckbox;
			btnSave.Visible = canSave;
			btnSave.Text = _node.BiometricType == NBiometricType.Voice ? "Save audio file" : "Save image";

			generalizeProgressView.PropertyChanged += GeneralizeProgressViewPropertyChanged;

			base.OnNavigatedTo(args);
		}

		public override void OnNavigatingFrom()
		{
			generalizeProgressView.PropertyChanged -= GeneralizeProgressViewPropertyChanged;

			panelView.Controls.Clear();
			IDisposable viewControl = _view as IDisposable;
			if (viewControl != null)
				viewControl.Dispose();
			_view = null;
			icaoWarningView.Face = null;
			generalizeProgressView.IcaoView = null;
			generalizeProgressView.View = null;

			base.OnNavigatingFrom();
		}

		#endregion

		#region Private static methods

		private static bool WasIcaoCheckPerformed(NFace face)
		{
			var attributes = face.Objects.FirstOrDefault();
			if (attributes != null && !attributes.TokenImageRect.IsEmpty)
			{
				return true;
			}
			else
			{
				var parentObject = (NLAttributes)face.ParentObject;
				return parentObject != null && !parentObject.TokenImageRect.IsEmpty;
			}
		}

		#endregion

		#region Private events

		private void GeneralizeProgressViewPropertyChanged(object sender, PropertyChangedEventArgs e)
		{
			if (e.PropertyName == "Selected")
			{
				if (_node.BiometricType == NBiometricType.Finger || _node.BiometricType == NBiometricType.Palm)
				{
					NFingerView view = (NFingerView)_view;
					NFrictionRidge finger = view.Finger;
					_biometric = finger;
					chbShowBinarized.Visible = finger.BinarizedImage != null;
					if (!chbShowBinarized.Visible && chbShowBinarized.Checked)
						chbShowBinarized.Checked = false;
				}
				else if (_node.BiometricType == NBiometricType.Face)
				{
					NFaceView view = (NFaceView)_view;
					_biometric = view.Face;
				}
			}
		}

		private void BtnFinishClick(object sender, EventArgs e)
		{
			PageController.NavigateToStartPage();
			panelView.Controls.Clear();
		}

		private void ChbShowBinarizedCheckedChanged(object sender, EventArgs e)
		{
			NFingerView view = _view as NFingerView;
			if (view != null)
			{
				view.ShownImage = chbShowBinarized.Checked ? ShownImage.Result : ShownImage.Original;
			}
		}

		private void BtnSaveClick(object sender, EventArgs e)
		{
			saveFileDialog.Filter = _biometric is NVoice ? string.Empty : NImages.GetSaveFileFilterString();
			saveFileDialog.FileName = string.Empty;
			if (saveFileDialog.ShowDialog() == DialogResult.OK)
			{
				try
				{
					if (_biometric.BiometricType == NBiometricType.Voice)
					{
						NVoice voice = _biometric as NVoice;
						voice.SoundBuffer.Save(saveFileDialog.FileName);
					}
					else
					{
						NImage image = null;
						if (_biometric.BiometricType == NBiometricType.Finger || _biometric.BiometricType == NBiometricType.Palm)
						{
							NFrictionRidge frictionRidge = (NFrictionRidge)_biometric;
							image = chbShowBinarized.Checked ? frictionRidge.BinarizedImage : frictionRidge.Image;
						}
						else if (_biometric.BiometricType == NBiometricType.Iris)
						{
							image = ((NIris)_biometric).Image;
						}
						else
						{
							image = ((NFace)_biometric).Image;
						}
						image.Save(saveFileDialog.FileName);
					}
				}
				catch (Exception ex)
				{
					Utilities.ShowError(ex);
				}
			}
		}

		#endregion
	}
}

File: java/Samples/Biometrics/VB/AbisSampleVB/SubjectEditor/BiometricPreviewPage.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Gui
Imports Neurotec.Images

Partial Public Class BiometricPreviewPage
	Inherits PageBase
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
		DoubleBuffered = True
	End Sub

#End Region

#Region "Private fields"

	Private _node As SubjectTreeControl.Node
	Private _biometric As NBiometric
	Private _view As Object

#End Region

#Region "Public methods"

	Public Overrides Sub OnNavigatedTo(ByVal ParamArray args() As Object)
		If args Is Nothing OrElse args.Length <> 1 Then
			Throw New ArgumentException("args")
		End If

		Dim showZoomControls As Boolean = True
		Dim showBinarizedCheckbox As Boolean = False
		Dim canSave As Boolean = False
		_view = Nothing
		_node = CType(args(0), SubjectTreeControl.Node)
		If _node Is Nothing OrElse (Not _node.IsBiometricNode) Then
			Throw New ArgumentException("args")
		End If
		panelView.Controls.Clear()
		GeneralizeProgressView.Clear()
		generalizeProgressView.Visible = _node.IsGeneralizedNode
		icaoWarningView.Visible = False
		_biometric = _node.Items.First()
		Select Case _node.BiometricType
			Case NBiometricType.Finger, NBiometricType.Palm
				Dim first As NFrictionRidge = TryCast(_biometric, NFrictionRidge)
				Dim view As New NFingerView() With {.Dock = DockStyle.Fill, .ForeColor = Color.Red, .Finger = first}
				_view = view
				panelView.Controls.Add(view)
				horizontalZoomSlider.View = view
				showBinarizedCheckbox = first.BinarizedImage IsNot Nothing
				canSave = first.Image IsNot Nothing
				If GeneralizeProgressView.Visible Then
					Dim generalized = _node.GetAllGeneralized()
					GeneralizeProgressView.View = view
					GeneralizeProgressView.Biometrics = _node.Items
					GeneralizeProgressView.Generalized = generalized
					GeneralizeProgressView.Selected = If(generalized.FirstOrDefault(), first)
				End If

				If showBinarizedCheckbox AndAlso chbShowBinarized.Checked Then
					view.ShownImage = ShownImage.Result
				End If
				Exit Select
			Case NBiometricType.Iris
				Dim view As New NIrisView() With {.Dock = DockStyle.Fill, .Iris = TryCast(_biometric, NIris)}
				panelView.Controls.Add(view)
				horizontalZoomSlider.View = view
				_view = view
				canSave = view.Iris.Image IsNot Nothing
				Exit Select
			Case NBiometricType.Face
				Dim first As NFace = TryCast(_biometric, NFace)
				Dim view As New NFaceView() With {.Dock = DockStyle.Fill, .Face = first, .ShowIcaoArrows = False}
				_view = view
				panelView.Controls.Add(view)
				horizontalZoomSlider.View = view
				canSave = view.Face.Image IsNot Nothing
				If _node.AllItems.Cast(Of NFace).Any(AddressOf WasIcaoCheckPerformed) Then
					icaoWarningView.Visible = True
				End If
				If GeneralizeProgressView.Visible Then
					Dim generalized = _node.GetAllGeneralized()
					generalizeProgressView.View = view
					generalizeProgressView.IcaoView = icaoWarningView
					GeneralizeProgressView.Biometrics = _node.Items
					GeneralizeProgressView.Generalized = generalized
					generalizeProgressView.Selected = If(generalized.FirstOrDefault(), first)
				Else
					icaoWarningView.Face = first
				End If
				Exit Select
			Case NBiometricType.Voice
				showZoomControls = False
				Dim view As New VoiceView() With {.Voice = TryCast(_biometric, NVoice)}
				panelView.Controls.Add(view)
				_view = view
				canSave = view.Voice.SoundBuffer IsNot Nothing
				Exit Select
		End Select

		horizontalZoomSlider.Visible = showZoomControls
		chbShowBinarized.Visible = showBinarizedCheckbox
		btnSave.Visible = canSave
		btnSave.Text = If(_node.BiometricType = NBiometricType.Voice, "Save audio file", "Save image")

		AddHandler GeneralizeProgressView.PropertyChanged, AddressOf GeneralizeProgressViewPropertyChanged

		MyBase.OnNavigatedTo(args)
	End Sub

	Public Overrides Sub OnNavigatingFrom()
		RemoveHandler generalizeProgressView.PropertyChanged, AddressOf GeneralizeProgressViewPropertyChanged

		panelView.Controls.Clear()
		Dim viewControl As IDisposable = TryCast(_view, IDisposable)
		If viewControl IsNot Nothing Then
			viewControl.Dispose()
		End If
		_view = Nothing
		icaoWarningView.Face = Nothing
		generalizeProgressView.IcaoView = Nothing
		generalizeProgressView.View = Nothing

		MyBase.OnNavigatingFrom()
	End Sub

#End Region

#Region "Private static methods"

	Private Shared Function WasIcaoCheckPerformed(ByVal face As NFace) As Boolean
		Dim attributes = face.Objects.FirstOrDefault()
		If attributes IsNot Nothing AndAlso (Not attributes.TokenImageRect.IsEmpty) Then
			Return True
		Else
			Dim parentObject = CType(face.ParentObject, NLAttributes)
			Return parentObject IsNot Nothing AndAlso Not parentObject.TokenImageRect.IsEmpty
		End If
	End Function

#End Region

#Region "Private events"

	Private Sub GeneralizeProgressViewPropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
		If e.PropertyName = "Selected" Then
			If _node.BiometricType = NBiometricType.Finger OrElse _node.BiometricType = NBiometricType.Palm Then
				Dim view As NFingerView = CType(_view, NFingerView)
				Dim finger As NFrictionRidge = view.Finger
				_biometric = finger
				chbShowBinarized.Visible = finger.BinarizedImage IsNot Nothing
				If (Not chbShowBinarized.Visible) AndAlso chbShowBinarized.Checked Then
					chbShowBinarized.Checked = False
				End If
			ElseIf _node.BiometricType = NBiometricType.Face Then
				Dim view As NFaceView = CType(_view, NFaceView)
				_biometric = view.Face
			End If
		End If
	End Sub

	Private Sub BtnFinishClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnFinish.Click
		PageController.NavigateToStartPage()
		panelView.Controls.Clear()
	End Sub

	Private Sub ChbShowBinarizedCheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chbShowBinarized.CheckedChanged
		Dim view As NFingerView = TryCast(_view, NFingerView)
		If view IsNot Nothing Then
			view.ShownImage = If(chbShowBinarized.Checked, ShownImage.Result, ShownImage.Original)
		End If
	End Sub

	Private Sub BtnSaveClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSave.Click
		saveFileDialog.Filter = If(TypeOf _biometric Is NVoice, String.Empty, NImages.GetSaveFileFilterString())
		saveFileDialog.FileName = String.Empty
		If saveFileDialog.ShowDialog() = DialogResult.OK Then
			Try
				If _biometric.BiometricType = NBiometricType.Voice Then
					Dim voice As NVoice = TryCast(_biometric, NVoice)
					voice.SoundBuffer.Save(saveFileDialog.FileName)
				Else
					Dim image As NImage = Nothing
					If _biometric.BiometricType = NBiometricType.Finger OrElse _biometric.BiometricType = NBiometricType.Palm Then
						Dim frictionRidge As NFrictionRidge = CType(_biometric, NFrictionRidge)
						image = If(chbShowBinarized.Checked, frictionRidge.BinarizedImage, frictionRidge.Image)
					ElseIf _biometric.BiometricType = NBiometricType.Iris Then
						image = (CType(_biometric, NIris)).Image
					Else
						image = (CType(_biometric, NFace)).Image
					End If
					image.Save(saveFileDialog.FileName)
				End If
			Catch ex As Exception
				Utilities.ShowError(ex)
			End Try
		End If
	End Sub

#End Region
End Class
